Support PBC in all three dimensions in radius_graph_pbc

emsunshine opened the pull request:
This PR allows `ocpmodels.common.utils.radius_graph_pbc` to properly compute edges across periodic boundaries in all three spatial dimensions. This allows the function to be applied more broadly (i.e. to bulks). The previous implementation ignored periodic boundary conditions in the `z` direction unless the cutoff was above the hard-coded vacuum of 20A.

However, even removing the hard-coded limit did not give the expected results. This is because the previous implementation based on `torch.meshgrid` did not list all desired unit cells. This procedure can be done easily with `torch.cartesian_prod`.

For slabs, it is more efficient to ignore periodic boundary conditions in the `z` direction (this is equivalent to assuming no edges between one slab and another). Because of this, the new implementation allows one to specify which directions should be treated periodically. Full PBC is implemented as the default, although for most OCP applications, `[True, True, False]` will be a more efficient way to achieve the same results as the previous implementation.

Before:
![radius_graph_pbc_before](https://user-images.githubusercontent.com/93541000/187981720-1240e14e-e8ff-4318-a4ac-6c4fb90f8f1e.PNG)
After:
![radius_graph_pbc_after](https://user-images.githubusercontent.com/93541000/187981747-99e3bdc4-9a02-4e74-b5ed-039306fbbdc2.PNG)

abhshkdz commented:
This is _awesome_, thanks for the PR!

With `pbc = [True, True, True]` and a 6 A cutoff (so not crossing over the vacuum layer along `z` for OC20 structures):
- how much slower is this than the existing implementation that ignores `z`?
- are the returned edges identical to the existing implementation for OC20-like structures? If they aren't, then we might want to set the default to `[True, True, False]` to not break backward-compatibility for OC20 trained models. Open to suggestions here.
 
And I think we'd want to set the PBC directions somewhere in the config, so that it's easily changeable depending on the data.

abhshkdz commented:
Also, could you please run the committed files through the black linter in the [pre-commit hook](https://github.com/Open-Catalyst-Project/ocp/blob/main/.pre-commit-config.yaml) (install using [`pre-commit install`](https://github.com/Open-Catalyst-Project/ocp/blob/main/INSTALL.md)) to make [circleci](https://app.circleci.com/pipelines/github/Open-Catalyst-Project/ocp/1099/workflows/e0f8881d-4138-41af-bbaa-514666a59800/jobs/1177) happy.

wood-b commented:
Thanks for addressing this issue Ethan! One thing to add to @abhshkdz comments, it would be great to have some formal unit tests. The tests could similarly compare radius_graph_pbc to pymatgen/AtomsToGraphs for a few relevant systems (e.g. adslab, bulk, molecule).  Happy to chat offline about test/edge cases we should include.

emsunshine commented:
Thank you both for the feedback! I agree that `[True, True, False]` is a reasonable default to replicate previous behavior; I have changed it to this for now. I'm not sure where exactly in the configs you would want this option. Perhaps in the model config? I think this function is only used inside of models and in the lbfgs optimizer.

This implementation is not necessarily guaranteed to list edges in exactly the same order as before. This may lead to slightly different graphs if `max_num_neighbors_threshold` is reached.

I have added a test based on the `atoms.json` structure in the `tests/preprocessing` directory. I agree that adding more tests is a good idea, especially for bulks where PBC is relevant in all 3 directions.

As far as speed difference, my initial timings did not show any significant difference, even when using `[True, True, True]`. A more thorough test is not a bad idea, though. At worst, the slowdown could be 3x if the unit cells "above" and "below" the slab are also considered as potential neighbors.

abhshkdz commented:
>  I'm not sure where exactly in the configs you would want this option. Perhaps in the model config? I think this function is only used inside of models and in the lbfgs optimizer.

Yeah in the model config probably makes the most sense. Let me think about it a bit more, I can give it a shot.

> This implementation is not necessarily guaranteed to list edges in exactly the same order as before. This may lead to slightly different graphs if max_num_neighbors_threshold is reached.

Makes sense, I'll test this for some models.

> I have added a test based on the atoms.json structure in the tests/preprocessing directory. I agree that adding more tests is a good idea, especially for bulks where PBC is relevant in all 3 directions.

The unit test looks good! If you already have some such toy bulk structures, maybe you could extend `atoms.json`?
